### PR TITLE
Chi 2033 clear signing ethereum tx functions

### PR DESCRIPTION
**Pull request by kartikCypherock**

### CODE REVIEW

**Comment by vbcypherock** [comment]
Task linked: [CHI-2001 Wallet Connect Firmware Enabler](https://app.clickup.com/t/37308523/CHI-2001)

**Comment by vbcypherock** [comment]
Task linked: [CHI-2033 Clear Signing Implementation for Ethereum transaction functions](https://app.clickup.com/t/37308523/CHI-2033)

File: common/coin_support/abi_decode.c [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+       )
+    {
+        return;
+    }
+
+    memcpy(
+            pConvertedBuffer,
+            pAbiTypeData + ABI_DYNAMIC_METADATA_OFFET_BE,
+            ABI_DYNAMIC_METADATA_SIZE_IN_BYTES
+          );
```
Comment (ujjwal-cyph) [inline common/coin_support/abi_decode.c:187]: self TODO: revisit

File: common/coin_support/abi_decode.c [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                                    uint8_t *pAbiTypeData,
+                                    uint8_t *pConvertedBuffer
+                                  );
+
+/* Global functions
+ *****************************************************************************/
+uint8_t ABI_DecodeDynamicValByType(
+                                    ABI_Type_e inputAbiType,
+                                    uint8_t *pAbiTypeData,
+                                    uint8_t *pAbiTypeDataBase,
```
Comment (ujjwal-cyph) [inline common/coin_support/abi_decode.c:78]: if it is okay, make these const.

File: common/coin_support/abi_decode.c [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+                                    uint32_t *pNumBytesReturned,
+                                    uint8_t **dpAbiDataPtr
+                                  )
+{
+    uint8_t returnCode = ABI_DECODE_BAD_ARGUMENT;
+    
+    if (
+        (NULL == pAbiTypeData)              ||
+        (NULL == pAbiTypeDataBase)          ||
+        (NULL == pNumBytesReturned)
```
Comment (ujjwal-cyph) [inline common/coin_support/abi_decode.c:88]: check `dpAbiDataPtr` for `NULL`

File: common/coin_support/abi_decode.c [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+             * pAbiTypeData points now to the 32-byte data denoting num of bytes
+             * in the dynamic input
+             */
+            uint32_t numBytesInData;
+            ABI_GetDynamicMetaData(
+                                    ABI_bytes_e,
+                                    pAbiTypeData,
+                                    ((uint8_t *)&(numBytesInData))
+                                  );
+            pAbiTypeData += 32;
```
Comment (ujjwal-cyph) [inline common/coin_support/abi_decode.c:118]: check for overflow in read

**Review by ujjwal-cyph — CHANGES_REQUESTED** [review]

File: common/coin_support/eth.c [hunk 5]
```diff
@@ -419,4 +475,217 @@ void eth_get_fee_string(eth_unsigned_txn *eth_unsigned_txn_ptr, char *fee_decima
...
+                             );
+            }
+            else if (ABI_uint256_e == pArgumentAbiType[currArgument])
+            {
+                ABI_Stringify(
+                                ABI_uint256_e,
+                                pCurrEthTxnPayload,
+                                0
+                             );
+            }
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply common/coin_support/eth.c:683]: solved

File: common/coin_support/eth.c [hunk 6]
```diff
@@ -419,4 +475,217 @@ void eth_get_fee_string(eth_unsigned_txn *eth_unsigned_txn_ptr, char *fee_decima
...
+    ABI_Type_e const *pArgumentAbiType;
+    
+    numArgsInFunction = ETH_DetectFunction(functionTag, &pArgumentAbiType);
+
+    if (
+        (NULL == pArgumentAbiType)          ||
+        (0 == numArgsInFunction)
+       )
+    {
+        return;
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply common/coin_support/eth.c:607]: fixed.
we will return `ETH_UTXN_FUNCTION_NOT_FOUND` if pArgumentAbiType == NULL.

File: common/coin_support/abi_decode.h [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+#define ABI_address_e_OFFSET_BE                     (ABI_uint256_e_SIZE_IN_BYTES - ABI_address_e_SIZE_IN_BYTES)
+
+#define ABI_DYNAMIC_METADATA_SIZE_IN_BYTES          (4)
+#define ABI_DYNAMIC_METADATA_OFFET_BE               (ABI_uint256_e_SIZE_IN_BYTES - ABI_DYNAMIC_METADATA_SIZE_IN_BYTES)
+
+#define ABI_DECODE_BAD_ARGUMENT                     (0x11)
+#define ABI_DECODE_PROCESS_COMPLETE                 (0xAA)
+/* Enums
+ *****************************************************************************/
+typedef enum
```
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply common/coin_support/abi_decode.h:35]:
```suggestion
/// Ref: https://docs.soliditylang.org/en/v0.8.17/abi-spec.html#types
typedef enum
```

File: common/coin_support/abi_decode.h [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+/* MACROS
+ *****************************************************************************/
+#define ABI_ELEMENT_SZ_IN_BYTES						(32)
+#define Abi_uint256_e_SZ_IN_BYTES					(32)
+
+#define Abi_address_e_SZ_IN_BYTES					(20)
+#define Abi_address_e_OFFSET_BE	(Abi_uint256_e_SZ_IN_BYTES - Abi_address_e_SZ_IN_BYTES)
+
+#define ABI_DYN_METADATA_SZ_IN_BYTES				(4)
```
Comment (ujjwal-cyph) [inline common/coin_support/abi_decode.h:25]: let's mention purpose of this. As in, if we are trying/expecting a practical upper limit or something similar.

File: common/coin_support/eth.c [hunk 9]
```diff
@@ -61,9 +61,65 @@
 #include "int-util.h"
 #include "logger.h"
 #include "utils.h"
+#include "abi_decode.h"
 
 static uint8_t rlp_encode_decimal(uint64_t dec, uint8_t offset, uint8_t *metadata);
 
+#define EVM_swap_TAG            (0x7c025200)
```
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply common/coin_support/eth.c:68]: Check once if this is a trusted source for reference

```suggestion
/// Ref: https://www.4byte.directory/signatures/?bytes4_signature=0x7c025200
#define EVM_swap_TAG            (0x7c025200)
```

File: common/coin_support/eth.c [hunk 10]
```diff
@@ -61,9 +61,63 @@
...
+    Abi_uint256_e,
+    Abi_uint256_e,
+    Abi_bytes_dynamic_e,
+    Abi_bytes_dynamic_e
+};
+
+const char *EVM_swap_Title = "Function: swap";
+const char *EVM_swap_Signature = "swap(address,(address,address,address,address,uint256,uint256,uint256),bytes,bytes)";
+
+#define EVM_uniswapV3Swap_TAG           (0xe449022e)
```
Comment (ujjwal-cyph) [inline common/coin_support/eth.c:87]: Check once if these references are okay to add.

```suggestion
/// Ref: https://www.4byte.directory/signatures/?bytes4_signature=0xe449022e
#define EVM_uniswapV3Swap_TAG           (0xe449022e)
```

File: common/coin_support/eth.c [hunk 11]
```diff
@@ -61,9 +61,63 @@
...
+{
+    Abi_uint256_e,
+    Abi_uint256_e,
+    Abi_uint256_array_dynamic_e
+};
+
+const char *EVM_uniswapV3Swap_Title = "Function: uniswapV3Swap";
+const char *EVM_uniswapV3Swap_Signature = "uniswapV3Swap(uint256,uint256,uint256[])";
+
+#define EVM_safeTransferFrom_TAG        (0x42842e0e)
```
Comment (ujjwal-cyph) [inline common/coin_support/eth.c:99]:
```suggestion
/// Ref: https://www.4byte.directory/signatures/?bytes4_signature=0x42842e0e
#define EVM_safeTransferFrom_TAG        (0x42842e0e)
```

File: common/coin_support/eth.c [hunk 12]
```diff
@@ -61,9 +61,63 @@
...
+{
+    Abi_address_e,
+    Abi_address_e,
+    Abi_uint256_e
+};
+
+const char *EVM_safeTransferFrom_Title = "Function: safeTransferFrom";
+const char *EVM_safeTransferFrom_Signature = "safeTransferFrom(address,address,uint256)";
+
+#define EVM_deposit_TAG         (0xd0e30db0)
```
Comment (ujjwal-cyph) [inline common/coin_support/eth.c:111]:
```suggestion
/// Ref: https://www.4byte.directory/signatures/?bytes4_signature=0xd0e30db0
#define EVM_deposit_TAG         (0xd0e30db0)
```

File: common/coin_support/abi_decode.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply common/coin_support/abi_decode.h]: added

File: common/coin_support/eth.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply common/coin_support/eth.c]: added

File: common/coin_support/eth.c (hunk 12 above)
Comment (kartikCypherock) [reply common/coin_support/eth.c]: added

File: common/coin_support/eth.c (hunk 11 above)
Comment (kartikCypherock) [reply common/coin_support/eth.c]: added

File: common/coin_support/eth.c (hunk 10 above)
Comment (kartikCypherock) [reply common/coin_support/eth.c]: added

File: common/coin_support/abi_decode.h (hunk 8 above)
Comment (kartikCypherock) [reply common/coin_support/abi_decode.h]: ok added a comment

**Pull request merged**